openedx/frontend-app-course-authoring
# [FC-0036] feat: New "Add Tags" widget

## Description

This PR add the new "Add Tags" widget. Which separates out staging tags when checking them and committing (saving) them. 

![add-tags-widget](https://github.com/openedx/frontend-app-course-authoring/assets/6829768/f976e9c3-5822-41bd-979f-adc8e91416fd)

## Supporting information

Related Tickets:
- Closes https://github.com/openedx/modular-learning/issues/188

## Testing instructions

1. Run you local dev stack.
1. If you don't already have sample taxonomy/tags data, follow the instructions in this repo to generate sample data: https://github.com/open-craft/taxonomy-sample-data
1. Navigate to a sample course and open the tags drawer by clicking on "Manage Tags"
1. Expand some taxonomies and click on the "Add a Tag" select input
1. Confirm the following:
    1. When checking tags in the dropdown, it keeps track of which tags are being newly added (staged). If you presse cancel, no changes are made. If press "Add Tags", the new tags are saved
    1. Newly added tags appear as chips in the add tags field, and can be removed by clicking on their X.
    1. Existing tags (already applied) that were added before the user clicks "Add Tags" do not appear as chips in the "Add Tags" widget.
    1. You can type into the widget to search the tags by keyword
    1. The functionality from https://github.com/openedx/frontend-app-course-authoring/pull/799 still exists. i.e matches are highlighted, parent tags are expanded but only if their child/grandchild matches the search, and a message is displayed if there are no matching tags.
    1. You can continue to type to search/filter tags even when there are chips displayed inside the field.
    1. Very long tag names wrap to the next line; they are not truncated.
    1. "Load More" is only shown if there are more pages of tags and is updated with the new style
    1. The styling of checkboxes of implicit tags has changed. Instead of a dash, it is now a checkmark (the transparency remains the same i.e. 40% opacity).
    1. The inline "Add" button is only shown if there is at least one newly added tag/chip visible.
    1. The second "Add Tags" button at the bottom of the dropdown is disabled until/unless there is at least one newly added tag/chip visible.
    1. The dropdown is shown when the "Add Tags" search box or the dropdown itself has focus. If the user clicks off the dropdown onto somewhere else on the tagging drawer, the dropdown will close but the chips will remain in the box and the tags won't be added until/unless the user clicks "Add Tags".
    1. If the user selects "cancel" on the dropdown menu, their selections (i.e. chips) are forgotten and the select menu is closed.
    1. Existing (previously saved) tags are shown as disabled checkmarks and cannot be removed using the checkboxes in the dropdown "Add Tags" widget. Instead, users click on the (X) next to them to remove them. But newly-added tags (that appear as chips) can be undone by un-checking them in the tag tree dropdown. This is the same as clicking the X on the chip.
---
Private-ref: [FAL-3643](https://tasks.opencraft.com/browse/FAL-3643)
